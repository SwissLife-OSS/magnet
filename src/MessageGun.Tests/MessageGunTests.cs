using System;
using System.Net;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using FluentAssertions;
using Moq;
using Newtonsoft.Json;
using Snapshooter.Xunit;
using Xunit;

namespace Magnet
{
    public class MessageGunTests
    {
        [Fact]
        public async Task Trigger_Success()
        {
            //arrange
            HttpRequestMessage sentRequest = null;
            IHttpClientFactory factory = CreateHttpClientFactoryMock(msg =>
            {
                sentRequest = msg;

            });
            var gun = new MessageGun(factory, new MessageGunOptions());

            //act
            await gun.TriggerAsync(MessageBuilder.New("Pdf", "UnitTest"), default);


            //assert
            sentRequest.Should().NotBeNull();
            MagnetMessage message = await GetMessageFromRequest(sentRequest);
            message.MatchSnapshot(o => o
                   .IgnoreField("Id")
                   .IgnoreField("ReceivedAt"));

        }

        private static async Task<MagnetMessage> GetMessageFromRequest(
            HttpRequestMessage sentRequest)
        {
            var json = await sentRequest.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<MagnetMessage>(json);
        }

        private IHttpClientFactory CreateHttpClientFactoryMock(
            Action<HttpRequestMessage> beforeSendRequest  )
        {
            Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> handler =
                (HttpRequestMessage message, CancellationToken token) =>
                {
                    beforeSendRequest(message);
                    return Task.FromResult(new HttpResponseMessage(HttpStatusCode.OK));
                };

            var client = new HttpClient(new DelegatingHandlerStub(handler))
            { BaseAddress = new Uri("http://foo") };
            var mock = new Mock<IHttpClientFactory>(MockBehavior.Strict);
            mock.Setup(_ => _.CreateClient(It.IsAny<string>())).Returns(client);
            return mock.Object;
        }


        public class DelegatingHandlerStub : DelegatingHandler
        {
            private readonly Func<HttpRequestMessage,
                CancellationToken, Task<HttpResponseMessage>> _handlerFunc;

            public DelegatingHandlerStub(
                Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> handlerFunc)
            {
                _handlerFunc = handlerFunc;
            }

            protected override Task<HttpResponseMessage> SendAsync(
                HttpRequestMessage request,
                CancellationToken cancellationToken)
            {
                return _handlerFunc(request, cancellationToken);
            }
        }
    }
}
