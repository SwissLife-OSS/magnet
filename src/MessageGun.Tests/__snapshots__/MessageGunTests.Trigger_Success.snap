{
  "Id": "acaf3e41-c2bf-4cf9-9f51-6377c12cf10e",
  "Type": "Pdf",
  "ReceivedAt": "2019-11-27T12:26:27.1578435Z",
  "From": null,
  "To": [],
  "Body": null,
  "Properties": {},
  "Provider": "UnitTest"
}
